document when atomic loads are guaranteed read-only

RalfJung opened the pull request:
Based on this [discussion in Zulip](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Can.20.60Atomic*.3A.3Aload.60.20perform.20a.20write).

The values for x86 and x86_64 are complete guesswork on my side, and I have no clue what the values might be for other architectures. I hope we can get the right people to chime in to gather the required information. :)

I'll update Miri to respect these rules once we have more data.

RalfJung committed:
document when atomic loads are guaranteed read-only

cuviper commented:
I'm going to put this in the API domain, since it's making guarantees.

@rustbot label -T-libs +t-libs-api
r? libs-api

Amanieu commented:
Some additions to the table:
aarch64 8 bytes[^1]
arm: 4 bytes[^2]
riscv64: 8 bytes
riscv32: 4 bytes

[^1]: ARMv8.4 raises this to 16 bytes, but I don't think we need to document this here. We're only guaranteeing minimums.
[^2]: Similarly, this is raised to 8 bytes with the LPAE extension.

RalfJung commented:
Cc @rust-lang/opsem  for help in filling out the table

RalfJung committed:
add ARM and RISC-V values

bjorn3 commented:
Can GCC guarantee this too? cc @antoyo

programmerjake commented on library/core/src/sync/atomic.rs:
```diff
@@ -79,6 +79,30 @@
+//! | `aarch64` | 8 bytes |
+//! | `riscv32` | 4 bytes |
+//! | `riscv64` | 8 bytes |
+//!
```
```suggestion
//! | `powerpc64le-unknown-linux-gnu` | 16 bytes |
//!
```

LLVM generates calls to `__atomic_load_16` which always uses the `lq` instruction -- GCC and LLVM disagree on the instruction being enabled by default.

RalfJung replied:
Is that really specific to the entire triple, or does this apply to all powerpc64le targets?

programmerjake replied:
i'm not 100% sure (since the PowerISA spec is somewhat ambiguous), but it should also apply to aix and linux kernel

RalfJung replied:
I'm confused. This is entirely a property of the codegen backend, right? For which sizes of atomic access does LLVM emit operations that work on read-only pages? This shouldn't depend on the OS/vendor part of the triple, right?

programmerjake replied:
it does because different OSes supply different libatomic implementations and supply different default cpu feature sets. e.g. MacOS requires x86-64-v2 (approximately)

programmerjake replied:
on Power, AIX requires many more cpu features than Linux.

RalfJung replied:
Can those features be opted-out via `-C target-cpu/target-feature`? The number we give here is the minimum that we guarantee for all configurations of the target.

programmerjake replied:
> Can those features be opted-out via `-C target-cpu/target-feature`?
yes, but all that does is make llvm call `__atomic_load_16` which, for linux, just does the atomic load using the `lq` instruction you just opted out of using. in fact, llvm has that feature disabled by default on linux (gcc has it enabled by default)

RalfJung replied:
Okay.

If AIX and maxOS require more than Linux, would it then be correct to list 16 as a lower bound for all powerpc64le targets?

programmerjake replied:
> If AIX and maxOS require more than Linux, would it then be correct to list 16 as a lower bound for all powerpc64le targets?

modern macOS doesn't support PowerISA and hasn't for a long time iirc, it was an example for requiring additional features on x86-64.

16 is the correct lower bound for Linux with glibc and (according to the PowerISA spec.) AIX, I'd have to check musl and freebsd.

programmerjake replied:
8 is the lower bound when clang is the compiler that libatomic or equivalent was built with, since clang doesn't enable the `lq` instruction by default.

antoyo commented:
> Can GCC guarantee this too? cc @antoyo

I'm not sure I have enough context/understanding of the situation.
Does Rust support atomic loads of bigger size (e.g. more than 8 bytes on x86_64)?
Isn't `__atomic_load` always guaranteed read-only?
Or are we talking about cases where Rust atomic loads use `__atomic_compare_exchange`?

bjorn3 commented:
On x86 LLVM emits cmpxchg8b for 64bit loads and on x86_64 cmpxchg16b for 128bit loads as far as I understand, which is considered a write (and thus causes SIGSEGV on read-only memory) even if it wouldn't actually change the value at the memory location in question. For values of at most the register width LLVM will use mov on x86/x86_64 which is guaranteed read-only. Does GCC behave the same?

antoyo commented:
> On x86 LLVM emits cmpxchg8b for 64bit loads and on x86_64 cmpxchg16b for 128bit loads as far as I understand, which is considered a write (and thus causes SIGSEGV on read-only memory) even if it wouldn't actually change the value at the memory location in question. For values of at most the register width LLVM will use mov on x86/x86_64 which is guaranteed read-only. Does GCC behave the same?

The [doc](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) mentions:

> 16-byte integral types are also allowed if ‘__int128’ (see [128-bit Integers](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fint128.html)) is supported by the architecture.

but trying to do so on my computer results in the following error:

> undefined reference to `__atomic_load_16'

and I read somewhere that 16-byte atomic operations always call a function in gcc, so not read-only.

I assume we would need to call `__atomic_always_lock_free` for every value in this table to make sure, but I'm not sure it would make sense that "always_lock_free" means "guaranteed read-only".

dtolnay commented:
> I'm going to put this in the API domain, since it's making guarantees.

@rust-lang/libs-api:

This is module-level documentation for `core::sync::atomic`. https://doc.rust-lang.org/1.72.0/core/sync/atomic/index.html

As I understand it, the new guarantee introduced here would make the following sound: we want a value that can be atomically loaded but the backing data might either be the read-write memory where interior mutable values ordinarily go, or could be readonly memory on certain architectures that have this new guarantee.

```rust
#[repr(transparent)]
pub struct AtomicallyLoadableU64(AtomicU64);

impl AtomicallyLoadableU64 {
    #[cfg(any(target_arch = "x86_64", target_arch = "aarch64", target_arch = "riscv64"))]
    pub fn from_readonly(readonly: &u64) -> &Self {
        unsafe { ... }
    }

    pub fn from_interior_mutable(interior_mutable: &AtomicU64) -> &Self {
        unsafe { ... }
    }

    pub fn load(&self, ordering: Ordering) -> u64 {
        self.0.load(ordering)
    }
}
```

Perhaps we'd be happy to delegate this to a T-opsem FCP? I'm not sure that I would have any relevant API insight on this. It seems fine to guarantee, if this is a stable property these architectures and our compiler backends can guarantee.

@rfcbot poll libs-api -- opsem makes the call

RalfJung committed:
add general powerpc64le bound

(some powerpc64le targets can guarantee more, but for now it doesn't seem worth separating by OS/vendor)

programmerjake commented on library/core/src/sync/atomic.rs:
```diff
@@ -92,13 +92,14 @@
+//! | `riscv64`     | 8 bytes |
+//! | `powerpc64le` | 8 bytes |
 //!
 //! Any atomic `load` on read-only memory larger than the given size are Undefined Behavior. For
```
I would argue that, rather than always being undefined behavior, they are instead sometimes undefined behavior depending on enabled/default target features and OS libraries. e.g. on `powerpc64le-unknown-linux-gnu` and/or when the `quadword-atomics` feature is enabled, 16-byte atomic loads are perfectly fine.

RalfJung replied:
Saying "it's not UB sometimes" is useless, we'd have to list the concrete guarantees we are making. I don't feel like it's worth the effort of bloating this table to a huge monster that covers every single combination of architecture, OS, and set of enabled target features...

Note that this table doesn't say 16-byte atomic loads are forbidden. It says 16-byte atomic loads *from read-only memory* are forbidden.

programmerjake replied:
> Saying "it's not UB sometimes" is useless, we'd have to list the concrete guarantees we are making.

the concrete guarantees are not explicitly specified, but mostly follow target features.

I think declaring everything not explicitly listed here as UB is taking it too far, because that means the following is UB even though most any reasonable person would expect it to be fine:
```rust
#[cfg(target_arch = "powerpc64le")]
#[cfg(target_feature = "quadword-atomics")]
pub fn f(v: &u128) -> u128 {
    unsafe { (*(v as *const _ as *const AtomicU128)).load(Ordering::Relaxed) }
}
```

digama0 replied:
I agree with Jacob that we shouldn't say this is UB if it is okay on some targets. It sounds like the situation is more like some combination of "target-dependent" and "unspecified", with some targets getting clearer upper and lower bounds. I think a better way to phrase it is that it is UB to use an atomic if the load lowers to a read-write atomic operation on read-only memory, and whether this actually happens is target-dependent with some specific cases being specified explicitly.

RalfJung replied:
I'm strongly opposed to saying things like "UB unless the backend does X". We don't make any guarantees about what the backend does unless documented otherwise, and this PR *is* that documentation. So such a statement doesn't actually make any more code well-defined, while at the same time encouraging people to look at what the backend does and depend on it as if it were a stable guarantee.

RalfJung replied:
We could say "for architectures not in the table, the maximum size for which atomic loads are read-only is not specified can cannot be relied upon". That's logically equivalent to saying that the threshold size is 0, which is logically equivalent to what the PR already says.

digama0 replied:
> I'm strongly opposed to saying things like "UB unless the backend does X".

I think it should be read more as "UB unless predicate P holds", where P is only loosely constrained and is otherwise target-dependent. It's not actually compiler-dependent (or at least it shouldn't be). The bounds on P are there to help people reason about the common targets, and everything else seems like a spec hole to me, although we may for pragmatic reasons hold off on defining it.

> IOW, to make your statement useful, we need a table of how atomic loads get lowered on each target, and we need to say that if for a given target there's no entry in the table then one may not make any assumptions about how it is lowered.

Yes, that is the intent. If there are no bounds specified, then the programmer is not allowed to assume it is not UB and hence can't rely on it. This is not the same as it actually being UB though, just like relying on unspecified layout details is not UB if you are able to guess correctly by "insider information". In this case, this would cover things like Jacob's `quadword-atomics` feature, until/unless rustc specifies this feature more precisely or adds it to the table.

RalfJung replied:
I guess we can use wording similar to what I used in the ABI-guarantees PR.
```
//! Atomics loads that are larger than this threshold (and *all* atomic loads on targets not listed
//! in the table) might still be read-only under certain conditions, but that is not a stable
//! guarantee and should not be relied upon.
```

RalfJung committed:
define 'read-only memory'

RalfJung committed:
wording

RalfJung commented on library/core/src/sync/atomic.rs:
```diff
@@ -79,6 +79,42 @@
+//! | `aarch64`     | 8 bytes |
+//! | `riscv32`     | 4 bytes |
+//! | `riscv64`     | 8 bytes |
+//! | `powerpc64`   | 8 bytes |
```
@programmerjake I have generalized this from powerpc64le to powerpc64 since that seems to be the target_arch name.

programmerjake replied:
those are actually different because they use different major revisions of the ABI (as well as being different endian).

RalfJung replied:
Yeah I knew about endianess. But do they all have readonly 8-byte atomic loads?

programmerjake replied:
yes, but some have 16-byte atomic loads by default too. In particular, `powerpc64le-unknown-linux-gnu` -- LLVM generates a call to `libatomic` which is just the 16-byte atomic load instruction `lq` plus some fences.

RalfJung replied:
Yeah the intention of this table is to find reliable lower bounds. We can make better promises for particular situations later, I'd like to start with a reasonable baseline.

RalfJung commented:
@BurntSushi @joshtriplett @m-ou-se there's an [rfcbot poll](https://github.com/rust-lang/rust/pull/115577#issuecomment-1710786618) waiting for your checkmark here. :)

RalfJung commented:
The [t-libs-api rfcbot poll](https://github.com/rust-lang/rust/pull/115577#issuecomment-1710786618) has now reached the usual threshold for team consensus (a strict majority of approvals with at most two votes still open). So I assume we can hand this over to t-opsem.

@rfcbot merge

saethlin commented on library/core/src/sync/atomic.rs:
```diff
@@ -79,6 +79,42 @@
+//! | `riscv64`     | 8 bytes |
+//! | `powerpc64`   | 8 bytes |
+//!
+//! Atomics loads that are larger than this threshold (and *all* atomic loads on targets not listed
```
```suggestion
//! Atomic loads that are larger than this threshold (and *all* atomic loads on targets not listed
```

taiki-e commented:
> other architectures

AFAIK[^0], baseline is:

- aarch64: 8 bytes [^1]
- arm: 4 bytes (v6+ other than v6m, pre-v6 linux/android, or v6m/pre-v6 with `atomic-32` target feature. on pre-v6, this is [**relaxed ordering only**](https://godbolt.org/z/6GEYsnnPP)) [^2]
- hexagon: 4 bytes [^3]
- loongarch64: 8 bytes
- mips: 4 bytes (r2+) [^7]
- mips32r6: 4 bytes
- mips64: 8 bytes
- mips64r6: 8 bytes
- powerpc: 4 bytes
- powerpc64: 8 bytes [^4]
- riscv32: 4 bytes (only if `a` or `force-atomics` target feature enabled) [^5]
- riscv64: 8 bytes (only if `a` or `force-atomics` target feature enabled) [^5] 
- s390x: 8 bytes [^6]
- sparc: 4 bytes
- sparc64: 8 bytes

As [previously said](https://github.com/rust-lang/miri/pull/2464#discussion_r939636130), I recommend guaranteeing this only for relaxed ordering.

On some architectures, atomic loads larger than the pointer width are also read-only, but whether it is used or not also depends on the compiler backend and its version. In particular, 128-bit atomics are not stable in `std` and support by LLVM is incomplete.
So, I recommend not guaranteeing this on atomic loads larger than the pointer width for now.

[^0]: Mainly based on investigation (and knowledge) when implementing [atomic-maybe-uninit](https://github.com/taiki-e/atomic-maybe-uninit), [portable-atomic](https://github.com/taiki-e/portable-atomic), and [atomic-memcpy](https://github.com/taiki-e/atomic-memcpy).
[^1]: 16 bytes on armv8.4+ (lvm target feature `lse2`). However, LLVM's lowering of the 128-bit atomic load in this case is [incorrect](https://github.com/llvm/llvm-project/issues/62652) (relaxed/acquire load is fine though). It was also [discussed to use other (non-read-only) instructions for seqcst load](https://reviews.llvm.org/D141429#4083953). It wasn't done, but the reason was completely unrelated to whether it was read-only.
[^2]: On v6+, other orderings are also read-only. If LPAE is supported this should be 8 bytes, but AFAIK, LLVM doesn't implement optimizations depending on whether LPAE is supported or not, so it's probably actually 4 bytes. In some cases the relaxed load is also lowered to libcall, but AFAIK there currently is no such a builtin target (since https://github.com/rust-lang/rust/pull/100621).
[^3]: AFAIK, this is probably actually 8 bytes, but we [currently do not provide 64-bit atomics on this platform](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/hexagon_unknown_linux_musl.rs#L6) 
[^4]: On power8+ (e.g., ppc64le) with LLVM 15+ or GCC, this is 16 bytes (llvm target feature `quadword-atomics`), but we [currently do not provide 128-bit atomics on this platform](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/powerpc64le_unknown_freebsd.rs#L7) and [not fully supported by LLVM (atomic_load is fine though)](https://github.com/taiki-e/portable-atomic/blob/4342eb56b2669fa99a4448a4496ba0dd933bb062/src/imp/atomic128/intrinsics.rs#L24-L26).
[^5]: If neither `a` nor `force-atomics` are enabled, [`atomic_*` libcalls will be generated](https://godbolt.org/z/97x1YYYoc), but I'm not sure that we can always safely assume that they *guarantee* this property. [riscv32imc-esp-espidf](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/riscv32imc_esp_espidf.rs) is the only builtin target that currently might have the problem.
[^6]: This is 16 bytes, but we [currently do not provide 128-bit atomics on this platform](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/s390x_unknown_linux_gnu.rs#L13) and [not fully supported by LLVM (atomic_load is fine though)](https://github.com/taiki-e/portable-atomic/blob/main/src/imp/atomic128/intrinsics.rs#L20-L23).
[^7]: The SYNC instruction was added in MIPSr2, so at least in MIPS(r1) the acquire/seqcst load should be lowered to libcalls as in ARM pre-v6. I don't think [LLVM currently supports those targets well enough](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/mipsel_sony_psx.rs#L24), though.

RalfJung commented:
So, in that entire table, every single row just says that the threshold is the pointer width. Do we really want to have a big table that repeats the pointer width in every single row?

RalfJung commented:
> If neither a nor force-atomics are enabled, [atomic_* libcalls will be generated](https://godbolt.org/z/97x1YYYoc), but I'm not sure that we can always safely assume that they guarantee this property. [riscv32imc-esp-espidf](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/riscv32imc_esp_espidf.rs) is the only builtin target that currently might have the problem.

Rust atomics generally guarantee that they are not implemented with a lock. So if these RISC-V features are missing, then our Atomic types should just not be available -- right?

taiki-e commented:
> So, in that entire table, every single row just says that the threshold is the pointer width. Do we really want to have a big table that repeats the pointer width in every single row?

With csky, recently supported in rustc (this target didn't exist when I said on this topic before in miri PR), LLVM seems to always generate libcalls.
https://github.com/llvm/llvm-project/commit/ec2de7490813a7593dad59f210c7ec41f1a29002

cc @Dirreke: Are csky-unknown-linux-gnuabiv2's `atomic_load_{1,2,4}` builtins guaranteed to work with read-only memory?
(Mentioned because you are [target maintainer](https://github.com/rust-lang/rust/blob/2d08657901d8fec8f51f203dc8a38e25c162e834/src/doc/rustc/src/platform-support/csky-unknown-linux-gnuabiv2.md#target-maintainers))

---

> > If neither a nor force-atomics are enabled, [atomic_* libcalls will be generated](https://godbolt.org/z/97x1YYYoc), but I'm not sure that we can always safely assume that they guarantee this property. [riscv32imc-esp-espidf](https://github.com/rust-lang/rust/blob/ab45885dec2a6552cb060a5b7183653baaecd580/compiler/rustc_target/src/spec/riscv32imc_esp_espidf.rs) is the only builtin target that currently might have the problem.
> 
> Rust atomics generally guarantee that they are not implemented with a lock. So if these RISC-V features are missing, then our Atomic types should just not be available -- right?
> 
> Looking at that target file they seem to enable atomics anyway. This is wrong unless we have a gaurantee that the builtins are implemented in a lock-free way.

I'm not 100% sure if I'm looking at the correct file, but it looks like:

- On single-core systems, it [disables interrupts](https://github.com/espressif/esp-idf/blob/3b748a6cb76c2db7c6368a0dea32a88bc58bc44d/components/newlib/stdatomic.c#L40) for all atomics. 
  - This doesn't use locks.
  - It seems [atomic_load_N is read-only](https://github.com/espressif/esp-idf/blob/3b748a6cb76c2db7c6368a0dea32a88bc58bc44d/components/newlib/stdatomic.c#L102).
- On multi-core systems, it seems to use [LL/SC for {8,16}-bit atomics (even if load/store)](https://github.com/espressif/esp-idf/blob/3b748a6cb76c2db7c6368a0dea32a88bc58bc44d/components/newlib/port/riscv/port_stdatomic.S), [global spinlocks for 64-bit atomics](https://github.com/espressif/esp-idf/blob/3b748a6cb76c2db7c6368a0dea32a88bc58bc44d/components/newlib/stdatomic.c#L64).
  - 64-bit atomic builtins are not lock-free -- Setting the max_atomic_width to 64 seems wrong.
  - Considering the memory ordering, {8,16}-bit atomic builtins seem wrong (it is always relaxed ordering)
  - Using LL/SC for load/store on riscv is also odd. Why not use `l{b,h,w}`/`s{b,h,w}` like [RISC-V Atomics ABI Specification](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/HEAD/riscv-atomic.adoc) or [RISC-V Instruction Set Manual](https://five-embeddev.com/riscv-isa-manual/latest/memory.html#sec:memory:porting)?
    - It seems that atomic_load_N uses only LL (`lr.w`), but I don't know if this is guaranteed to work correctly on riscv's read-only memory.

I think this would be fine if riscv32imc-esp-espidf is *guaranteed* to be used only on single-core systems, but atomic builtins handle the difference, so I guess it probably actually works on multi-core systems.

cc @ivmarkov @MabezDev
(Mentioned because you both are [target maintainers](https://github.com/rust-lang/rust/blob/2d08657901d8fec8f51f203dc8a38e25c162e834/src/doc/rustc/src/platform-support/esp-idf.md#target-maintainers))

---

> And if they are lock-free, they should work on read-only memory, no?

I don't think the two are correlated. For example, atomic loads using CAS are lock-free, but do write.

Dirreke commented:
> > So, in that entire table, every single row just says that the threshold is the pointer width. Do we really want to have a big table that repeats the pointer width in every single row?
> 
> With csky, recently supported in rustc (this target didn't exist when I said on this topic before in miri PR), LLVM seems to always generate libcalls. [llvm/llvm-project@ec2de74](https://github.com/llvm/llvm-project/commit/ec2de7490813a7593dad59f210c7ec41f1a29002)
> 
> cc @Dirreke: Are csky-unknown-linux-gnuabiv2's `atomic_load_{1,2,4}` builtins guaranteed to work with read-only memory? (Mentioned because you are [target maintainer](https://github.com/rust-lang/rust/blob/2d08657901d8fec8f51f203dc8a38e25c162e834/src/doc/rustc/src/platform-support/csky-unknown-linux-gnuabiv2.md#target-maintainers))

Actually I have no idea. Which file should I read to figure it out? does LLVM do this?

MabezDev commented:
> Considering the memory ordering, {8,16}-bit atomic builtins seem wrong (it is always relaxed ordering)
Using LL/SC for load/store on riscv is also odd. Why not use l{b,h,w}/s{b,h,w} like [RISC-V Atomics ABI Specification](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/HEAD/riscv-atomic.adoc) or [RISC-V Instruction Set Manual](https://five-embeddev.com/riscv-isa-manual/latest/memory.html#sec:memory:porting)?
It seems that atomic_load_N uses only LL (lr.w), but I don't know if this is guaranteed to work correctly on riscv's read-only memory.

Thank you for pointing that out, I will direct that feedback to the relevant folks at Espressif.

> I think this would be fine if riscv32imc-esp-espidf is guaranteed to be used only on single-core systems, but atomic builtins handle the difference, so I guess it probably actually works on multi-core systems.

This was true at the time of implementation and is true now, but won't be soon with the release of the [ESP32-P4](https://www.espressif.com/en/news/ESP32-P4), which is a dual-core RISCV  chip with only 32-bit atomics. We are already running into this with our Xtensa targets in the [esp-rs/rust fork](https://github.com/esp-rs/rust), something that was brought to my attention by @workingjubilee in https://github.com/esp-rs/rust/pull/196 just yesterday.

It seems for arm linux there are builtins emitted by [compiler_builtins](https://github.com/rust-lang/compiler-builtins/blob/master/src/arm_linux.rs) to support older arm chips without CAS. The documentation for [`__kuser_cmpxchg`](https://www.kernel.org/doc/Documentation/arm/kernel_user_helpers.txt) doesn't appear to guarantee lock-free behaviour, in fact it makes a note that the internal implementation can change at any time, only the interface is stable.

We don't want to lose 64-bit atomics for these esp-idf STD targets. In practice, it works. How should we go about resolving this?

Amanieu commented:
> As [previously said](https://github.com/rust-lang/miri/pull/2464#discussion_r939636130), I recommend guaranteeing this only for relaxed ordering.

I know of real use cases for load-acquire on read-only memory. Specifically when working with shared memory to which you have a read-only view but which can be mutated by another process.

taiki-e commented:
@Amanieu

> > As [previously said](https://github.com/rust-lang/miri/pull/2464#discussion_r939636130), I recommend guaranteeing this only for relaxed ordering.
> 
> I know of real use cases for load-acquire on read-only memory. Specifically when working with shared memory to which you have a read-only view but which can be mutated by another process.

The problem is that load-acquire calls CAS on some targets ([godbolt](https://godbolt.org/z/6GEYsnnPP)). For pre-v6 ARM Linux/Android it is possible to adjust load-acquire to call `LDR` and `__kuser_memory_barrier` (like https://github.com/taiki-e/portable-atomic/pull/112) on our part, but for other pre-v6 ARM and pre-r2 MIPS, I'm not sure it is possible to resolve that issue.

Also, I think you can use `load(Relaxed)` with `fence(Acquire)` instead of `load(Acquire)`. Doesn't it work for your use case?

---

@Dirreke

> > > So, in that entire table, every single row just says that the threshold is the pointer width. Do we really want to have a big table that repeats the pointer width in every single row?
> > 
> > 
> > With csky, recently supported in rustc (this target didn't exist when I said on this topic before in miri PR), LLVM seems to always generate libcalls. [llvm/llvm-project@ec2de74](https://github.com/llvm/llvm-project/commit/ec2de7490813a7593dad59f210c7ec41f1a29002)
> > cc @Dirreke: Are csky-unknown-linux-gnuabiv2's `atomic_load_{1,2,4}` builtins guaranteed to work with read-only memory? (Mentioned because you are [target maintainer](https://github.com/rust-lang/rust/blob/2d08657901d8fec8f51f203dc8a38e25c162e834/src/doc/rustc/src/platform-support/csky-unknown-linux-gnuabiv2.md#target-maintainers))
> 
> Actually I have no idea. Which file should I read to figure it out? does LLVM do this?

I have no idea. I guess the atomic builtins are provided by libatomic or others, because nothing seems to be implemented in LLVM. (LLVM just generates the code that calls atomic builtins)

---

@MabezDev

> We don't want to lose 64-bit atomics for these esp-idf STD targets. In practice, it works. How should we go about resolving this?

I think the reason it works in your cases is that you have not run code that depends on std 64-bit atomics being lock-free on a multi-core system. I wouldn't be surprised if something suddenly stops working once you get to a situation that meets that condition.

Well, it is not rare for 64-bit atomics to be unavailable on 32-bit systems. AFAIK, usually on such a target, Atomic*64, is avoided, or a [lock-based fallback is defined](https://github.com/tokio-rs/tokio/blob/707fb4d0df72d7ee6a4f4b5a675f84cd1863bc94/tokio/src/loom/std/atomic_u64_as_mutex.rs), or a [crate that provides something like that is used](https://github.com/metrics-rs/metrics/blob/67d5100f6d4c1d52efcc1cc5c3b1987e83ebffd8/metrics/Cargo.toml#L29-L30).

RalfJung commented:
> I don't think the two are correlated. For example, atomic loads using CAS are lock-free, but do write.

Right, okay, so libcalls can be lock-free. We just need to make sure they are, for all calls emitted by Rust's standard atomic types. Or, for the purpose of this PR -- libcalls can be read-only sometimes. We just need to make sure they are, for all relaxed loads that are no bigger than a pointer size.

Dirreke commented:
> Dirreke
> 
> > > > So, in that entire table, every single row just says that the threshold is the pointer width. Do we really want to have a big table that repeats the pointer width in every single row?
> > > 
> > > 
> > > With csky, recently supported in rustc (this target didn't exist when I said on this topic before in miri PR), LLVM seems to always generate libcalls. [llvm/llvm-project@ec2de74](https://github.com/llvm/llvm-project/commit/ec2de7490813a7593dad59f210c7ec41f1a29002)
> > > cc @Dirreke: Are csky-unknown-linux-gnuabiv2's `atomic_load_{1,2,4}` builtins guaranteed to work with read-only memory? (Mentioned because you are [target maintainer](https://github.com/rust-lang/rust/blob/2d08657901d8fec8f51f203dc8a38e25c162e834/src/doc/rustc/src/platform-support/csky-unknown-linux-gnuabiv2.md#target-maintainers))
> > 
> > 
> > Actually I have no idea. Which file should I read to figure it out? does LLVM do this?
> 
> I have no idea. I guess the atomic builtins are provided by libatomic or others, because nothing seems to be implemented in LLVM. (LLVM just generates the code that calls atomic builtins)

I'm not sure if I'm right. csky-unknown-linux-gnueabiv2 will link to `libatomic.a`( at [/rustc_target/spec/csky-unknown-linux-gnuabiv2.rs](https://github.com/rust-lang/rust/blob/c7224e3c954ce54853364c950608c1f0ed47ea85/compiler/rustc_target/src/spec/csky_unknown_linux_gnuabiv2.rs#L15) ) which is provided by gcc 6.3. It's a gcc builtin function at [gcc/builtins.c](https://github.com/c-sky/gcc/blob/7dfc9fd6f0b3d9ba256ca36ee6bc512f741ae8d7/gcc/builtins.c#L5187-L5209). Is it can said that it work with read-only memory or not?

RalfJung commented:
The `expand_atomic_load` that this dispatches to has [this comment](https://github.com/c-sky/gcc/blob/7dfc9fd6f0b3d9ba256ca36ee6bc512f741ae8d7/gcc/optabs.c#L6293-L6295) so it seems like loads until and including the "word size" (hopefully the same as "pointer size") are just regular reads?

taiki-e commented:
@Dirreke

> csky-unknown-linux-gnueabiv2 will link to `libatomic.a`( at [csky-unknown-linux-gnuabiv2.rs](https://github.com/rust-lang/rust/blob/c7224e3c954ce54853364c950608c1f0ed47ea85/compiler/rustc_target/src/spec/csky_unknown_linux_gnuabiv2.rs#L15) ) which is provided by gcc 6.3.

Thanks. I downloaded the toolchain mentioned in the [platform support doc](https://github.com/rust-lang/rust/blob/b6ab01a7131481b52131a9bd814dc8084052f643/src/doc/rustc/src/platform-support/csky-unknown-linux-gnuabiv2.md) and checked the assemblies contained in `libatomic.a` included therein.

There are multiple `libatomic.a` and two patterns of atomic_load_{1,2,4} implementation were found:

csky-linux-gnuabiv2/lib/ck800/libatomic.a is:

```
ld.{b,h,w}      	r0, (r0, 0x0)
```


other csky-linux-gnuabiv2/lib/*/libatomic.a and csky-linux-gnuabiv2/lib/libatomic.a are:

```
sync
ld.{b,h,w}      	r0, (r0, 0x0)
sync
```

These seem to be implemented using read-only instructions. (Given that older CPUs seem to have no SYNC instruction,  in the future LLVM may lower acquire/seqcst loads in the same way as pre-v6 ARM (i.e. call __sync_val_compare_and_swap_N for acquire/seqcst loads), though.)

However, apart from that, the atomic RMW implementation provided in libatomic.a does not seem to be lock-free.

<details/><summary/>details</summary>

The following two are implementations in libatomic.a of the __atomic_exchange_4 builtin called in 32-bit atomic swap. It seems that the assembly calls a function to acquire the critical section, performs the operation, and then releases the critical section.

csky-linux-gnuabiv2/lib/ck800/libatomic.a is:

```
00000000 <__atomic_exchange_4>:
   0:	14d3      	push      	r4-r6, r15
   2:	6d03      	mov      	r4, r0
   4:	6d87      	mov      	r6, r1
   6:	e0000000 	bsr      	0x0	// 0 <libat_lock_1>
   a:	94a0      	ld.w      	r5, (r4, 0x0)
   c:	6c13      	mov      	r0, r4
   e:	b4c0      	st.w      	r6, (r4, 0x0)
  10:	e0000000 	bsr      	0x0	// 0 <libat_unlock_1>
  14:	6c17      	mov      	r0, r5
  16:	1493      	pop      	r4-r6, r15
```

csky-linux-gnuabiv2/lib/libatomic.a is:

```
00000000 <__atomic_exchange_4>:
   0:	14d3      	push      	r4-r6, r15
   2:	6d03      	mov      	r4, r0
   4:	6d87      	mov      	r6, r1
   6:	eae00007 	jsri      	0x0	// from address pool at 0x20
   a:	c4004820 	lsli      	r0, r0, 0
   e:	94a0      	ld.w      	r5, (r4, 0x0)
  10:	6c13      	mov      	r0, r4
  12:	b4c0      	st.w      	r6, (r4, 0x0)
  14:	eae00004 	jsri      	0x0	// from address pool at 0x24
  18:	c4004820 	lsli      	r0, r0, 0
  1c:	6c17      	mov      	r0, r5
  1e:	1493      	pop      	r4-r6, r15
```


It is not clear what function is being called in lib/libatomic.a, but the function being called at lib/ck800/libatomic.a is calling pthread_mutex_{un}lock internally.

```
00000000 <libat_lock_1>:
   0:	14d0      	push      	r15
   2:	e4002fc0 	andi      	r0, r0, 4032
   6:	1161      	lrw      	r3, 0x0	// 88 <libat_unlock_n+0x34>
   8:	600c      	addu      	r0, r3
   a:	e0000000 	bsr      	0x0	// 0 <pthread_mutex_lock>
   e:	1490      	pop      	r15

00000010 <libat_unlock_1>:
  10:	14d0      	push      	r15
  12:	e4002fc0 	andi      	r0, r0, 4032
  16:	107d      	lrw      	r3, 0x0	// 88 <libat_unlock_n+0x34>
  18:	600c      	addu      	r0, r3
  1a:	e0000000 	bsr      	0x0	// 0 <pthread_mutex_unlock>
  1e:	1490      	pop      	r15
```

It seems that [csky has LL/SC](https://github.com/torvalds/linux/blob/3aba70aed91f2b283f7952be152ad76ec5c34975/arch/csky/include/asm/atomic.h#L31), but are those not always available, even on CPUs with SYNC instruction?



</details>

Dirreke commented:
> It seems that [csky has LL/SC](https://github.com/torvalds/linux/blob/3aba70aed91f2b283f7952be152ad76ec5c34975/arch/csky/include/asm/atomic.h#L31), but are those not always available, even on CPUs with SYNC instruction?

Very grateful for doing this. I checked the ISA user guide and it said that ldex/stex is supported on all 32bit target. So I think it's always avalable.
![image](https://github.com/rust-lang/rust/assets/51513111/e437d570-a86b-4cef-98e8-0f51a9983b10)

programmerjake commented:
on gcc on rv32i without a, all atomic loads generate just standard loads and fences (no libcalls), which are part of the base rv32i
https://godbolt.org/z/jKo6nc9q1

taiki-e commented:
> on gcc on rv32i without a, all atomic loads generate just standard loads and fences (no libcalls), which are part of the base rv32i

Interesting. It seems similar to the behavior of old LLVM on ARM targets that do not support CAS (https://github.com/rust-lang/rust/issues/99668, https://reviews.llvm.org/D120026). However, trying to use RMW is worse because it calls atomic builtins that may be incompatible with load/store ([godbolt](https://godbolt.org/z/Woo15cK3T)). (It is not a problem when using in Rust, though.)

RalfJung commented:
So, that sounds like we need an issue tracking that on that target, our atomics aren't lock-free as they should be? I haven't really been able to understand the details; @taiki-e I'd appreciate if you could file an issue. :)

But for this PR, the claim that ptr-sized-and-smaller relaxed atomic loads are read-only seems to be confirmed?

programmerjake commented:
> @taiki-e I'd like to get your take on this -- do you think we should have a per-target table or just say "ptr sized" (since according to your table, this does match the pointer size for all our targets)?

I think a per-target table is a good idea, since some targets support *more* than pointer-sized, e.g. powerpc64le-unknown-linux-gnu supports 128-bit. and rv32i on current llvm using llvm's version of libatomic seems to not support lock-free atomics at all?

RalfJung commented:
The table we had didn't list more than ptr size for any target, and taiki-e's table didn't either, so for now I think the table would be entirely identical to just saying "ptr-sized".

Rust atomics must always be lock-free so that rv32i case sounds like a separate issue.

programmerjake commented:
> The table we had didn't list more than ptr size for any target,

well, what can I say, I wanted to add powerpc64le-unknown-linux-gnu, but others insisted that each arch must only have one entry...

workingjubilee commented:
Well if the default is usize atomicity, but some are atomic for larger, then we can have a match expression with a fallthrough case. :^)

taiki-e commented:
@RalfJung 

> @taiki-e I'd like to get your take on this -- do you think we should have a per-target table or just say "ptr sized" (since according to your table, this does match the pointer size for all our targets)?

The ones I know are okay at this time are the architectures mentioned in https://github.com/rust-lang/rust/pull/115577#issuecomment-1732204636, x86, x86_64, csky, and avr. msp430[^1] and m68k[^2] are probably okay too, but I'm not sure about WASM, BPF, and GPUs.

Looking at the [GCC code you linked](https://github.com/rust-lang/rust/pull/115577#issuecomment-1732592646), it seems to be okay to say "ptr sized", but if we want to make conclusions without looking into the details about them, it would be safer to use a per-target table.

---

@programmerjake

> I think a per-target table is a good idea, since some targets support _more_ than pointer-sized, e.g. powerpc64le-unknown-linux-gnu supports 128-bit. and rv32i on current llvm using llvm's version of libatomic seems to not support atomics at all?

As for ppc64le, as mentioned in https://github.com/rust-lang/rust/pull/115577#issuecomment-1732204636, `core` doesn't provide 128-bit atomic for it, and due to LLVM bug, it will be difficult to provide it right now. s390x is similar.

As for riscv without A extension, only riscv32imc-esp-espidf currently provides atomic loads.

- There are [some bugs and oddities](https://github.com/rust-lang/rust/pull/115577#issuecomment-1732259297) in ESP-IDF's atomic builtins, but the instructions used there seem to work with read-only memory as far as I could tell from a simple test with QEMU. (I'm not certain that the ISA guarantees that, though.)
- Except for ESP-IDF, atomic is not provided at all in the first place. https://github.com/rust-lang/rust/pull/114499 will fix this, and it will not violate the guarantee here if it is merged because it enables `forced-atomics` feature (generate load/store instructions instead of calling atomic builtins).

[^1]: LLVM (and `core`) does not support atomic on this target, but if implemented correctly the load will be read-only.
[^2]: https://reviews.llvm.org/D136525

programmerjake commented:
> As for ppc64le, as mentioned in [#115577 (comment)](https://github.com/rust-lang/rust/pull/115577#issuecomment-1732204636), `core` doesn't provide 128-bit atomic for it, and due to LLVM bug, it will be difficult to provide it right now.

well, for powerpc64le-unknown-linux-gnu specifically, llvm currently compiles 16-byte atomics to calls to libatomic, *however* that's still perfectly fine since those 16-byte libatomic functions *are lock free* (they use the quadword atomics instructions since libatomic is compiled by gcc and gcc enables quadword atomics by default). the atomic load uses `lq` which works on read-only memory.

taiki-e commented:
> lvm currently compiles 16-byte atomics to calls to libatomic

No. At least since LLVM 15 (https://github.com/llvm/llvm-project/commit/549e118e93c666914a1045fde38a2cac33e1e445), LLVM generates 16-byte atomic instructions instead of calling libatomic for powerpc64le-unknown-linux-gnu (https://godbolt.org/z/3rebKcbdf). What I said was problematic is that when I try to use certain kinds of atomic operations, LLVM crashes (https://godbolt.org/z/MjvfYMT63), or the generated code is wrong.

programmerjake commented:
> > llvm currently compiles 16-byte atomics to calls to libatomic
> 
> No. At least since LLVM 15 ([llvm/llvm-project@549e118](https://github.com/llvm/llvm-project/commit/549e118e93c666914a1045fde38a2cac33e1e445)), LLVM generates 16-byte atomic instructions instead of calling libatomic for powerpc64le-unknown-linux-gnu (https://godbolt.org/z/3rebKcbdf). What I said was problematic is that when I try to use certain atomic operations, LLVM crashes (https://godbolt.org/z/MjvfYMT63), or the generated code is wrong ([when I checked before](https://github.com/taiki-e/portable-atomic/commit/ce1b2514a2fa9ad6147855da2f649f7d5d11dc81), at least the fetch_add was broken).

well, all the atomics I tried all call libatomic when compiling the llvm ir with recent clang, so maybe clang enables the special Make The Atomics Work™ flag that we just have to find? The llvm ir from that LLVM crash link above just calls libatomic with clang 17.0.1:
https://llvm.godbolt.org/z/jYsfEd8ae

taiki-e commented:
> well, all the atomics I tried all call libatomic when compiling the llvm ir with recent clang, so maybe clang enables the special Make The Atomics Work™ flag that we just have to find? The llvm ir from that LLVM crash link above just calls libatomic with clang 17.0.1:
> https://llvm.godbolt.org/z/jYsfEd8ae

In Clang, powerpc64le-unknown-linux-gnu does not seem to imply that quadword-atomics is enabled.
The crash can be reproduced by passing -mcpu=pwr8.
https://llvm.godbolt.org/z/4zdPnKcbT

programmerjake commented:
I opened https://github.com/llvm/llvm-project/issues/68390 for the atomic max ICE

RalfJung commented:
Eh, I can't be reviewing my own PR.^^

r? libs-api 
I guess? Or lang?

RalfJung commented:
Hm okay bouncing back to the same reviewer is no good.^^
? @Amanieu 

Also I'll need to incorporate some feedback still.
@rustbot author

RalfJung commented:
Hello, bot?
? @Amanieu

stepancheg commented:
“Undefined behavior” is a bit too strong. UB means the compiler may use it in optimizations to reformat your disk. Which is not the case here IIUC. The worst that can happen is your program crashes. More correct wording might be “unspecified behavior”.

digama0 commented:
I don't think so. If you try to write to read-only memory, the compiler is allowed to optimize assuming you didn't, it is not required to issue a write that causes some target-dependent behavior. This is what we usually mean by undefined behavior, and while formatting your hard drive is an unlikely consequence it is difficult to put sensible bounds on what can happen when optimizations are used outside their realm of applicability which is why we generally don't try.

programmerjake commented:
> I don't think so. If you try to write to read-only memory, the compiler is allowed to optimize assuming you didn't, it is not required to issue a write that causes some target-dependent behavior.

I do not expect any reasonable compiler to treat an atomic load as writing and optimize based off that...

digama0 commented:
I think the issue is the reverse: a compiler might assume an atomic load is a load, which does not cause a fault and does not induce data races with other reading threads. These things are not true if the load is simulated using compare-exchange, and the resulting data race can cause other issues.

programmerjake commented:
> I think the issue is the reverse: a compiler might assume an atomic load is a load, which does not cause a fault and does not induce data races with other reading threads. These things are not true if the load is simulated using compare-exchange, and the resulting data race can cause other issues.

I think that compilers can generally consider all atomic loads, including ones that are compiled to compare-exchange or similar, to not data-race with other reads, because the written data is always the exact same as the read data, so it's as if the write never occurred from the compiler IR perspective. the only remaining issues the write causes is a seg-fault if the page is mapped as read-only, possible issues with writing to a memory-mapped device (you probably need volatile for that anyway), and the effects on caching (which at worst only make the program run slower, it's still correct).

RalfJung commented:
Yeah this is definitely UB, it can cause worse things than a crash. Specifically, something that you wrote in code after the atomic load, might be moved by the compiler to happen before the atomic load. So you were expecting a crash but you're getting arbitrary effects from code after the load.

This is very similar to how e.g. an out-of-bounds volatile load/store is UB.

RalfJung committed:
only guarantee for Relaxed; add ptr-size fallback

RalfJung commented:
All right, I have brought back the table. It lists exceptions for powerpc64le and s390x (though that's kind of moot since we don't support 128bit atomics on those targets anyway). I didn't like the fallback idea since when people add a new target, they are not going to be aware that they need to check this table.

@taiki-e @programmerjake can you double-check the final form of this PR just to be sure? :)
@Amanieu this is ready for review
@rustbot ready

programmerjake commented on library/core/src/sync/atomic.rs:
```diff
@@ -79,6 +79,39 @@
+//! system via platform-specific APIs.
+//!
+//! As an exception from the general rule stated above, "sufficiently small" atomic loads with
+//! `Ordering::Relaxed` are implemented in a way that works on read-only memory, and are hence not
```
maybe mention that, if you need an acquire load, you can use a relaxed load immediately followed by an acquire fence. there might be something for sequentially consistent too, but i'm not sure.

RalfJung replied:
Good point, will do.

I don't think there's anything like that for SC. SC accesses are *weird*. (SC fences are perfectly reasonable though.)

RalfJung committed:
acquire loads can be done as relaxed load; acquire fence

taiki-e commented on library/core/src/sync/atomic.rs:
```diff
@@ -79,6 +79,42 @@
+//!
+//! | Target triple prefix (regular expression) | Size limit |
+//! |---------------|---------|
+//! | `i(3|5|6)86-`, `arm`, `thumb`, `mips(|el)-`, `powerpc-`, `riscv32`, `sparc-`    | 4 bytes |
+//! | `x86_64-`, `aarch64-`, `loongarch64-`, `mips64(|el)-`, `powerpc64-`, `riscv64`  | 8 bytes |
+//! | `powerpc64le-` | 16 bytes |
+//! | `s390x-`       | 16 bytes |
```
```suggestion
//! | `i(3|5|6)86-`, `arm`, `thumb`, `mips(|el)-`, `mipsisa32r6(|el)-`, `powerpc-`, `riscv32`, `sparc-`    | 4 bytes |
//! | `x86_64`, `aarch64`, `arm64`, `loongarch64-`, `mips64(|el)-`, `mipsisa64r6(|el)-`, `powerpc64-`, `riscv64`, `sparc(64|v9)-`  | 8 bytes |
//! | `powerpc64le-`, `s390x-`       | 16 bytes |
```

Changes in the above suggestion:
- x86_64- -> x86_64: x86_64h(-apple-darwin) is not x86_64- but is x86_64
- aarch64- -> aarch64: aarch64_be-(-unknown-linux-gnu(|_ilp32)|-unknown-linux-netbsd) are not aarch64- but is aarch64
- add arm64: arm64_32(-apple-watchos) is not aarch64- but is aarch64
- add sparc(64|v9)- (target_arch: sparc64), mipsisa32r6(|el)- (target_arch: mips32r6), mipsisa64r6(|el)- (target_arch: mips64r6)
- merge powerpc64le- and s390x- into the same row

Since target triples are complex and cannot be handled correctly on the user side without a build script, I recommend using target_arch based tables if possible.

RalfJung replied:
> I recommend using target_arch based table (like https://github.com/rust-lang/rust/pull/115577#issuecomment-1732204636) if possible.

Yeah that's what we had earlier, but powerpc64le- has target_arch = powerpc64 so then we can't tell it apart from powerpc64-.

Would it make sense to use target_cpu?

taiki-e replied:
My recommendation is to guarantee nothing about 128-bit atomics as said in https://github.com/rust-lang/rust/pull/115577#issuecomment-1732204636.

RalfJung replied:
Well I guess we're back to a table that just says "pointer size" is a convoluted way then. :shrug:

RalfJung committed:
use target-arch based table

taiki-e commented on library/core/src/sync/atomic.rs:
```diff
@@ -79,6 +79,41 @@
+//! |---------------|---------|
+//! | `x86`, `arm`, `mips`, `mips32r6, `powerpc`, `riscv32`, `sparc`, `hexagon` | 4 bytes |
+//! | `x86_64`, `aarch64`, `loongarch64`, `mips64`, `mips64r6`, `powerpc64`, `riscv64`, `sparc64` | 8 bytes |
+//! | `s390x`, `powerpc64` with `target_feature = "quadword-atomics"` | 16 bytes |
```
quadword-atomics is not a valid rust target feature. Also note that even if valid, the powerpc target feature is not stable.

https://github.com/rust-lang/rust/blob/49691b1f70d71dd7b8349c332b7f277ee527bf08/compiler/rustc_codegen_ssa/src/target_features.rs#L225-L235

taiki-e replied:
I think it is reasonable to document it once the 128-bit atomics in that target and the corresponding target feature have stabilized.

RalfJung committed:
remove 128bit atomics, they are anyway not exposed on those targets

RalfJung commented:
All right, so we have FCP passed and approval from our domain expert.
@Amanieu I think this is ready to land. :)

Amanieu commented:
I'm still not very happy about not supporting `load(Acquire)`, I honestly think it's a bug that this doesn't work on read-only memory and we should fix targets (like armv5) where this happens. But this can be done later.

@bors r+ rollup

bors commented:
:pushpin: Commit e494df436df105f8a3d767dfcd8e51c597c0fca5 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu e494df436df105f8a3d767dfcd8e51c597c0fca5 -->
<!-- homu: {"type":"Approved","sha":"e494df436df105f8a3d767dfcd8e51c597c0fca5","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-timer commented:
Finished benchmarking commit (93e62a260f9edb5813a73908120a43f1415f2c8c): [comparison URL](https://perf.rust-lang.org/compare.html?start=616e37919c87f34d3af57ab7457186a4e0cd62ef&end=93e62a260f9edb5813a73908120a43f1415f2c8c&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=616e37919c87f34d3af57ab7457186a4e0cd62ef&end=93e62a260f9edb5813a73908120a43f1415f2c8c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=616e37919c87f34d3af57ab7457186a4e0cd62ef&end=93e62a260f9edb5813a73908120a43f1415f2c8c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.7%  | [1.7%, 1.7%]   | 1     |
| Improvements ✅ <br /> (primary)   | -3.2% | [-3.2%, -3.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.9% | [-2.8%, -0.9%] | 2     |
| All ❌✅ (primary)                 | -3.2% | [-3.2%, -3.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=616e37919c87f34d3af57ab7457186a4e0cd62ef&end=93e62a260f9edb5813a73908120a43f1415f2c8c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.6% | [-5.6%, -5.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=616e37919c87f34d3af57ab7457186a4e0cd62ef&end=93e62a260f9edb5813a73908120a43f1415f2c8c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 626.338s -> 623.96s (-0.38%)
**Artifact size**: 305.62 MiB -> 305.57 MiB (-0.02%)

taiki-e commented:
@RalfJung

> So, that sounds like we need an issue tracking that on that target, our atomics aren't lock-free as they should be? I haven't really been able to understand the details; @taiki-e I'd appreciate if you could file an issue. :)

Filed as #117305 and #117306.
